src/economy/shipping_schedule.cc

```diff
@@ -0,0 +1,1715 @@
+}
+
+void ShippingSchedule::port_added(Game& game, PortDock& dock) {
+	sslog("\nShippingSchedule::port_added (%u)\n", dock.serial());
+	if (fleet_.count_ports() > 1) {
+		// nothing to do currently
+		sslog("nothing to do\n\n");
+		return;
+	}
+	// All ships are most likely panicking because they have
+	// no destination. Send them all to the new port.
+	for (Ship* ship : fleet_.get_ships()) {
+		assert(!ship->get_destination());
+		sslog("Rerouting %s there\n", ship->get_shipname().c_str());
+		ship->set_destination(game, &dock);
+		for (ShippingItem& si : ship->items_) {
+			assert(!si.destination_dock_.is_set());
+			si.destination_dock_ = &dock;
+		}
+	}
+	sslog("--- port_added maintenance complete ---\n\n");
+}
+
+Duration ShippingSchedule::update(Game& game) {
+	/*
+	 * This function is the heart of the shipping system.
+	 * All decisions (except emergency decisions on port destruction) are made here.
+	 * Here, we decide which ship will when arrive at which port and how many items
+	 * for which destinations it will pick up.
+	 * When we were notified that a ship or port was added or lost, we do some
+	 * maintenance around this fact in other functions, but our main job is to check
+	 * on every call every single port whether it has wares that will not be
+	 * transported anytime soon. If so, we can tell a nearby idle ship to pick up
+	 * wares, or order a ship that is already heading there to pick them up (if it
+	 * makes sense), or leave them for now for lack of capacity. We may also cancel
+	 * coming ships if transfers were cancelled so we need less capacity than
+	 * previously planned.
+	 * Additionally, we will distribute idle ships more or less equally among ports
+	 * so that every port will have a ship at hand immediately when it requires one
+	 * (provided that we have enough ships, of course).
+	 * In theory, it would be good to implement intelligent reordering of ships'
+	 * destinations, so as to avoid routes like A-B-C where A and C are close and B
+	 * is far away. We used to have such an algorithm, but it was shown to
+	 * necessarily be a performance killer (I only say Travelling Salesman Problem),
+	 * so we dropped support for this approach and instead prefer the GOLDEN RULE that
+	 * a ship should never service too many destinations at once, REGARDLESS of their
+	 * relative distances. One ship from B to A plus one ship from B to C are better
+	 * than one ship from B to both A and C. Instead we prefer to distribute tasks
+	 * among many ships. This produces the best results when the player builds a large
+	 * naval force. (When the player has few ships for many ports, this approach will
+	 * work suboptimally, but that is a bad strategy so the player deserves no more.)
+	 */
+
+	sslog("\nShippingSchedule::update\n");
+	assert(plans_.size() == fleet_.get_ships().size());
+	if (fleet_.get_ships().empty()) {
+		sslog("No ships\n");
+		return endless();  // wait until we get a ship
```

Comment:
How will the retriggering happen?

Reply:
Whenever a new ware/worker arrives at a port it will call fleet_->update() which will enqueue a retriggering of this function.